jobsdb/unionQueryLegacy.go

```diff
@@ -5,20 +5,62 @@ type MultiTenantLegacy struct {
 }
 
 func (mj *MultiTenantLegacy) GetAllJobs(workspaceCount map[string]int, params GetQueryParamsT, _ int) []*JobT {
+	var list []*JobT
 	toQuery := 0
 	for workspace := range workspaceCount {
 		toQuery += workspaceCount[workspace]
 	}
-	retryList := mj.GetToRetry(GetQueryParamsT{CustomValFilters: params.CustomValFilters, JobCount: toQuery})
-	toQuery -= len(retryList)
-	waitList := mj.GetWaiting(GetQueryParamsT{CustomValFilters: params.CustomValFilters, JobCount: toQuery})
-	toQuery -= len(waitList)
-	unprocessedList := mj.GetUnprocessed(GetQueryParamsT{CustomValFilters: params.CustomValFilters, JobCount: toQuery})
+	params.JobsLimit = toQuery
+	eventsLimitEnabled := params.EventsLimit > 0
+	payloadLimitEnabled := params.PayloadSizeLimit > 0
 
-	var list []*JobT
+	retryList := mj.GetToRetry(params)
 	list = append(list, retryList...)
+	updateParams(eventsLimitEnabled, payloadLimitEnabled, &params, retryList)
+	if limitsReached(eventsLimitEnabled, payloadLimitEnabled, params) {
+		return list
+	}
+
+	waitList := mj.GetWaiting(params)
 	list = append(list, waitList...)
-	list = append(list, unprocessedList...)
+	updateParams(eventsLimitEnabled, payloadLimitEnabled, &params, waitList)
+	if limitsReached(eventsLimitEnabled, payloadLimitEnabled, params) {
+		return list
+	}
 
+	unprocessedList := mj.GetUnprocessed(params)
+	list = append(list, unprocessedList...)
 	return list
 }
+
+func updateParams(eventsLimitEnabled, payloadLimitEnabled bool, params *GetQueryParamsT, jobs []*JobT) {
+	params.JobsLimit -= len(jobs)
+	if eventsLimitEnabled {
+		params.EventsLimit -= getTotalEvents(jobs)
+	}
+	if payloadLimitEnabled {
+		params.PayloadSizeLimit -= getTotalPayload(jobs)
+	}
+}
+
+func limitsReached(eventsLimitEnabled, payloadLimitEnabled bool, params GetQueryParamsT) bool {
+	return params.JobsLimit <= 0 ||
+		(eventsLimitEnabled && params.EventsLimit <= 0) ||
+		(payloadLimitEnabled && params.PayloadSizeLimit <= 0)
+}
+
+func getTotalEvents(jobs []*JobT) int {
```

Comment:
Do we need this check since we call GetAllJobs only in router and every job is an event